File: app/src/main/java/starbright/com/projectegg/features/ingredients/IngredientsFragment.java
package starbright.com.projectegg.features.ingredients;

import android.Manifest;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.os.Environment;
import android.provider.MediaStore;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.content.FileProvider;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.PopupMenu;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import com.afollestad.materialdialogs.MaterialDialog;
import com.google.firebase.auth.FirebaseAuth;

import java.io.File;
import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.List;
import java.util.Locale;
import java.util.Timer;
import java.util.TimerTask;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import id.zelory.compressor.Compressor;
import permissions.dispatcher.NeedsPermission;
import permissions.dispatcher.RuntimePermissions;
import starbright.com.projectegg.MyApp;
import starbright.com.projectegg.R;
import starbright.com.projectegg.data.AppRepository;
import starbright.com.projectegg.data.local.model.Ingredient;
import starbright.com.projectegg.util.ClarifaiHelper;
import starbright.com.projectegg.util.Constants;
import starbright.com.projectegg.util.scheduler.BaseSchedulerProvider;

import static android.app.Activity.RESULT_OK;

@RuntimePermissions
public class IngredientsFragment extends Fragment implements IngredientsContract.View,
        IngredientsAdapter.Listener {

    private static final int AUTOCOMPLETE_DELAY = 600;
    private static final int CAMERA_REQUEST_CODE = 101;

    private static final String BUNDLE_CART = "BUNDLE_CART";

    @Inject
    AppRepository repository;

    @Inject
    Compressor compressor;

    @Inject
    BaseSchedulerProvider schedulerProvider;

    @BindView(R.id.et_search_ingredients)
    EditText etSearchIngredients;

    @BindView(R.id.img_action_button)
    ImageView imgActionButton;

    @BindView(R.id.rv_ingredients)
    RecyclerView rvIngredients;

    @BindView(R.id.suggestion_progress_bar)
    ProgressBar suggestionProgressBar;

    @BindView(R.id.tv_cart_count)
    TextView tvCartCount;

    private Timer mTimer;

    private IngredientsAdapter mSearchSuggestionAdapter;
    private IngredientsContract.Presenter mPresenter;
    private MaterialDialog mDialog;
    private CartBottomSheetDialogFragment mCartBottomSheetDialogFragment;
    private String mCurrentPhotoPath;

    private FragmentListener mFragmentListener;

    private TextWatcher mIngredientsTextWatcher = new TextWatcher() {
        @Override
        public void beforeTextChanged(CharSequence s, int start, int count, int after) {

        }

        @Override
        public void onTextChanged(CharSequence s, int start, int before, int count) {
            if (mTimer != null) {
                mTimer.cancel();
            }
            mPresenter.handleSuggestionTextChanged(s.toString());
        }

        @Override
        public void afterTextChanged(final Editable s) {
            mTimer = new Timer();
            mTimer.schedule(new TimerTask() {
                @Override
                public void run() {
                    if (getActivity() != null) {
                        getActivity().runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                mPresenter.searchIngredient(s.toString());
                            }
                        });
                    }
                }
            }, AUTOCOMPLETE_DELAY);
        }
    };

    public static IngredientsFragment newInstance() {
        return new IngredientsFragment();
    }

    @Override
    public void onAttach(Context context) {
        MyApp.getAppComponent().inject(this);
        super.onAttach(context);
        mFragmentListener = (FragmentListener) context;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        new IngredientsPresenter(
                repository,
                this,
                schedulerProvider,
                new ClarifaiHelper(getActivity()),
                compressor
        );
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater,
                             @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        final View view = inflater.inflate(R.layout.fragment_ingredients, container, false);
        ButterKnife.bind(this, view);
        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        mPresenter.start();
        if (savedInstanceState != null) {
            mPresenter.setCart(savedInstanceState.<Ingredient>getParcelableArrayList(BUNDLE_CART));
        }
    }

    @Override
    public void onDetach() {
        mFragmentListener = null;
        super.onDetach();
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (resultCode == RESULT_OK && requestCode == CAMERA_REQUEST_CODE) {
            mPresenter.handleCameraResult(mCurrentPhotoPath);
        }
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putParcelableArrayList(BUNDLE_CART, mPresenter.getCart());
    }

    @Override
    public void setPresenter(IngredientsPresenter presenter) {
        mPresenter = presenter;
    }

    @Override
    public void setupEtSearchIngredients() {
        etSearchIngredients.addTextChangedListener(mIngredientsTextWatcher);
    }

    @Override
    public void setupRvIngredientSuggestion() {
        final LinearLayoutManager layoutManager = new LinearLayoutManager(getActivity(),
                LinearLayoutManager.VERTICAL, false);
        rvIngredients.setLayoutManager(layoutManager);

        mSearchSuggestionAdapter = new IngredientsAdapter(getActivity());
        mSearchSuggestionAdapter.setListener(this);
        rvIngredients.setAdapter(mSearchSuggestionAdapter);
    }

    @Override
    public void setupMaterialProgressDialog() {
        mDialog = new MaterialDialog.Builder(getActivity())
                .title(R.string.ingredients_dialog_progress_title)
                .content(R.string.ingredients_dialog_progress_content)
                .progress(true, 0)
                .canceledOnTouchOutside(false)
                .build();
    }

    @Override
    public void setupBottomSheetDialogFragment() {
        mCartBottomSheetDialogFragment = new CartBottomSheetDialogFragment();
        mCartBottomSheetDialogFragment.setListener(new CartBottomSheetDialogFragment.SheetListener() {
            @Override
            public void updateCart(List<Ingredient> ingredients) {
                mPresenter.setCart(ingredients);
            }

            @Override
            public void submitButtonClicked() {
                mFragmentListener.navigateRecipeListActivity(mPresenter.getCart());
            }
        });
    }

    @Override
    public void clearEtSearchQuery() {
        etSearchIngredients.setText("");
    }

    @Override
    public void openCamera() {
        IngredientsFragmentPermissionsDispatcher.startCameraIntentWithPermissionCheck(this);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode,
                                           @NonNull String[] permissions,
                                           @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        IngredientsFragmentPermissionsDispatcher.onRequestPermissionsResult(this,
                requestCode, grantResults);
    }

    @Override
    public void showSearchSuggestion(List<Ingredient> ingredients) {
        rvIngredients.setVisibility(View.VISIBLE);
        mSearchSuggestionAdapter.setIngredients(ingredients);
    }

    @Override
    public void showActionCamera() {
        imgActionButton.setVisibility(View.VISIBLE);
        imgActionButton.setImageResource(R.drawable.ic_camera);
    }

    @Override
    public void showActionClear() {
        imgActionButton.setVisibility(View.VISIBLE);
        imgActionButton.setImageResource(R.drawable.ic_clear);
    }

    @Override
    public void hideActionButton() {
        imgActionButton.setVisibility(View.GONE);
    }

    @Override
    public void showSuggestionProgressBar() {
        suggestionProgressBar.setVisibility(View.VISIBLE);
    }

    @Override
    public void hideSuggestionProgressBar() {
        suggestionProgressBar.setVisibility(View.GONE);
    }

    @Override
    public void showItemEmptyToast() {
        Toast.makeText(getActivity(), getString(R.string.ingredients_search_empty),
                Toast.LENGTH_SHORT).show();
    }

    @Override
    public void showItemMaxToast() {
        Toast.makeText(getActivity(), getString(R.string.ingredients_cart_error_max),
                Toast.LENGTH_SHORT).show();
    }

    @Override
    public void showDuplicateItemToast() {
        Toast.makeText(getActivity(), getString(R.string.ingredients_cart_error_duplicate),
                Toast.LENGTH_SHORT).show();
    }

    @Override
    public void showPredictionEmptyToast() {
        Toast.makeText(getActivity(), getString(R.string.ingredients_prediction_empty),
                Toast.LENGTH_SHORT).show();
    }

    @Override
    public void showItemAddedToast() {
        Toast.makeText(getActivity(), getString(R.string.ingredients_added_text),
                Toast.LENGTH_LONG).show();
    }

    @Override
    public void showSuccessPutIngredientToast(String ingredientName) {
        Toast.makeText(getActivity(), getString(R.string.ingredients_cart_included_format,
                ingredientName), Toast.LENGTH_SHORT).show();
    }

    @Override
    public void showMaterialProgressDialog() {
        mDialog.show();
    }

    @Override
    public void hideMaterialProgressDialog() {
        if (mDialog != null) {
            mDialog.dismiss();
        }
    }

    @Override
    public void updateIngredientCount(int count) {
        tvCartCount.setText(String.valueOf(count));
    }

    @Override
    public void hideSoftKeyboard() {
        final InputMethodManager imm = (InputMethodManager) getActivity()
                .getSystemService(Activity.INPUT_METHOD_SERVICE);
        if (imm != null) {
            imm.toggleSoftInput(InputMethodManager.HIDE_IMPLICIT_ONLY, 0);
        }
    }

    @Override
    public void showBottomSheetDialog() {
        mCartBottomSheetDialogFragment.show(getChildFragmentManager(),
                mCartBottomSheetDialogFragment.getTag());
    }

    @Override
    public void setCartItem(List<Ingredient> cart) {
        mCartBottomSheetDialogFragment.setCartIngredient(cart);
    }

    @Override
    public void hideSearchSuggestion() {
        rvIngredients.setVisibility(View.GONE);
    }

    @Override
    public void onSuggestionItemClicked(Ingredient selectedItem) {
        mPresenter.handleSuggestionItemClicked(selectedItem);
    }

    @NeedsPermission(Manifest.permission.CAMERA)
    public void startCameraIntent() {
        File photoFile = null;
        try {
            photoFile = createImageFile();
        } catch (IOException e) {
            e.printStackTrace();
        }
        if (photoFile != null) {
            Uri photoURI = FileProvider.getUriForFile(
                    getActivity(),
                    Constants.PROVIDER_PACKAGE_NAME,
                    photoFile);
            Intent intent = new Intent(MediaStore.ACTION_IMAGE_CAPTURE);
            intent.putExtra(MediaStore.EXTRA_OUTPUT, photoURI);
            startActivityForResult(intent, CAMERA_REQUEST_CODE);
        }
    }

    @OnClick({
            R.id.img_action_button,
            R.id.tv_cart_count,
            R.id.iv_more
    })
    void onClickedEvent(View view) {
        switch (view.getId()) {
            case R.id.img_action_button:
                mPresenter.handleActionButtonClicked(etSearchIngredients.getText().toString());
                break;
            case R.id.tv_cart_count:
                mPresenter.handleCartTvClicked();
                break;
            case R.id.iv_more:
                showPopupMenu(view);
                break;
        }
    }

    private File createImageFile() throws IOException {
        final String timeStamp = new SimpleDateFormat(Constants.YYYY_MM_DD_FORMAT, Locale.US)
                .format(new Date());

        final File storageDir = getActivity().getExternalFilesDir(Environment.DIRECTORY_PICTURES);
        final File image = File.createTempFile(timeStamp, Constants.JPG, storageDir);

        mCurrentPhotoPath = image.getAbsolutePath();
        return image;
    }

    private void showPopupMenu(View view) {
        final PopupMenu popup = new PopupMenu(getActivity(), view);
        final MenuInflater inflater = popup.getMenuInflater();
        inflater.inflate(R.menu.search_ingredient_menu, popup.getMenu());
        popup.setOnMenuItemClickListener(new PopupMenu.OnMenuItemClickListener() {
            @Override
            public boolean onMenuItemClick(MenuItem item) {
                switch (item.getItemId()) {
                    case R.id.menu_signout:
                        FirebaseAuth.getInstance().signOut();
                        mFragmentListener.navigateUserAccountActivity();
                        return true;
                    default:
                        return false;
                }
            }
        });
        popup.show();
    }

    interface FragmentListener {
        void navigateUserAccountActivity();

        void navigateRecipeListActivity(List<Ingredient> ingredients);
    }
}

File: app/src/main/java/starbright/com/projectegg/features/ingredients/IngredientsFragment.kt
/**
 * Created by Andreas on 11/8/2019.
 */

/**
 * Created by Andreas on 9/9/2018.
 */

package starbright.com.projectegg.features.ingredients

import android.Manifest
import android.annotation.SuppressLint
import android.app.Activity
import android.app.Activity.RESULT_OK
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.os.Environment
import android.provider.MediaStore
import android.support.v4.app.Fragment
import android.support.v4.content.FileProvider
import android.support.v7.widget.LinearLayoutManager
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.InputMethodManager
import android.widget.PopupMenu
import android.widget.Toast
import com.afollestad.materialdialogs.MaterialDialog
import com.google.firebase.auth.FirebaseAuth
import kotlinx.android.synthetic.main.fragment_ingredients.*
import permissions.dispatcher.NeedsPermission
import permissions.dispatcher.RuntimePermissions
import starbright.com.projectegg.R
import starbright.com.projectegg.data.model.Ingredient
import starbright.com.projectegg.util.Constants
import java.io.File
import java.io.IOException
import java.text.SimpleDateFormat
import java.util.*
import javax.inject.Inject
import kotlin.concurrent.schedule

@RuntimePermissions
class IngredientsFragment : Fragment(), IngredientsContract.View {

    @Inject
    lateinit var presenter: IngredientsPresenter

    private lateinit var cartBottomSheetDialogFragment: CartBottomSheetDialogFragment
    private lateinit var searchSuggestionAdapter: IngredientsAdapter
    private lateinit var dialog: MaterialDialog

    private var fragmentListener: FragmentListener? = null
    private var currentPhotoPath: String? = null
    private var timer: Timer? = null

    private val mIngredientsTextWatcher = object : TextWatcher {
        override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {

        }

        override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
            timer?.cancel()
            presenter.handleSuggestionTextChanged(s.toString())
        }

        override fun afterTextChanged(s: Editable) {
            timer = Timer().also {
                it.schedule(AUTOCOMPLETE_DELAY.toLong()) {
                    activity?.runOnUiThread { presenter.searchIngredient(s.toString()) }
                }
            }
        }
    }

    override fun onAttach(context: Context?) {
//        MyApp.getAppComponent().inject(this)
        super.onAttach(context)
        fragmentListener = context as FragmentListener?
    }

    @SuppressLint("NeedOnRequestPermissionsResult")
    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        onRequestPermissionsResult(requestCode, grantResults)
    }

    override fun onCreateView(inflater: LayoutInflater,
                              container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_ingredients, container, false)
        presenter.view = this
        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        presenter.start()
    }

    override fun onDetach() {
        fragmentListener = null
        et_search_ingredients.removeTextChangedListener(mIngredientsTextWatcher)
        cartBottomSheetDialogFragment.sheetListener = null
        presenter.onDestroy()
        super.onDetach()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (resultCode == RESULT_OK && requestCode == CAMERA_REQUEST_CODE) {
            presenter.handleCameraResult(currentPhotoPath!!)
        }
    }

    override fun setupEtSearchIngredients() {
        et_search_ingredients.addTextChangedListener(mIngredientsTextWatcher)
    }

    override fun setupRvIngredientSuggestion() {
        activity?.let {
            searchSuggestionAdapter = IngredientsAdapter(it).also { adapter ->
                adapter.onClickListener = { selectedIngredient ->
                    presenter.handleSuggestionItemClicked(selectedIngredient)
                }
            }
        }

        rv_ingredients.apply {
            layoutManager = LinearLayoutManager(activity, LinearLayoutManager.VERTICAL, false)
            adapter = searchSuggestionAdapter
        }
    }

    override fun setupImageActionButton() {
        img_action_button.setOnClickListener {
            presenter.handleActionButtonClicked(et_search_ingredients.text.toString())
        }
    }

    override fun setupTvCartCount() {
        tv_cart_count.setOnClickListener {
            presenter.handleCartTvClicked()
        }
    }

    override fun setupMoreButton() {
        iv_more.setOnClickListener {
            showPopupMenu(it)
        }
    }

    override fun setupMaterialProgressDialog() {
        dialog = MaterialDialog.Builder(activity as Context)
                .title(R.string.ingredients_dialog_progress_title)
                .content(R.string.ingredients_dialog_progress_content)
                .progress(true, 0)
                .canceledOnTouchOutside(false)
                .build()
    }

    override fun setupBottomSheetDialogFragment() {
        cartBottomSheetDialogFragment = CartBottomSheetDialogFragment().also {
            it.sheetListener = object : CartBottomSheetDialogFragment.SheetListener {
                override fun updateCart(ingredients: MutableList<Ingredient>) {
                    presenter.cart = ingredients.toMutableList()
                }

                override fun submitButtonClicked() {
                    fragmentListener?.navigateRecipeListActivity(presenter.cart)
                }
            }
        }
    }

    override fun clearEtSearchQuery() {
        et_search_ingredients.setText("")
    }

    override fun openCamera() {
        startCameraIntentWithPermissionCheck()
    }

    override fun showSearchSuggestion(ingredients: List<Ingredient>) {
        rv_ingredients.visibility = View.VISIBLE
        searchSuggestionAdapter.setIngredients(ingredients)
    }

    override fun showActionCamera() {
        img_action_button.apply {
            visibility = View.VISIBLE
            setImageResource(R.drawable.ic_camera)
        }
    }

    override fun showActionClear() {
        img_action_button.apply {
            visibility = View.VISIBLE
            setImageResource(R.drawable.ic_clear)
        }
    }

    override fun hideActionButton() {
        img_action_button.visibility = View.GONE
    }

    override fun showSuggestionProgressBar() {
        suggestion_progress_bar.visibility = View.VISIBLE
    }

    override fun hideSuggestionProgressBar() {
        suggestion_progress_bar.visibility = View.GONE
    }

    override fun showItemEmptyToast() {
        Toast.makeText(activity, getString(R.string.ingredients_search_empty),
                Toast.LENGTH_SHORT).show()
    }

    override fun showItemMaxToast() {
        Toast.makeText(activity, getString(R.string.ingredients_cart_error_max),
                Toast.LENGTH_SHORT).show()
    }

    override fun showDuplicateItemToast() {
        Toast.makeText(activity, getString(R.string.ingredients_cart_error_duplicate),
                Toast.LENGTH_SHORT).show()
    }

    override fun showPredictionEmptyToast() {
        Toast.makeText(activity, getString(R.string.ingredients_prediction_empty),
                Toast.LENGTH_SHORT).show()
    }

    override fun showItemAddedToast() {
        Toast.makeText(activity, getString(R.string.ingredients_added_text),
                Toast.LENGTH_LONG).show()
    }

    override fun showSuccessPutIngredientToast(ingredientName: String) {
        Toast.makeText(activity, getString(R.string.ingredients_cart_included_format,
                ingredientName), Toast.LENGTH_SHORT).show()
    }

    override fun showMaterialProgressDialog() {
        dialog.show()
    }

    override fun hideMaterialProgressDialog() {
        dialog.dismiss()
    }

    override fun updateIngredientCount(count: Int) {
        tv_cart_count.text = count.toString()
    }

    override fun hideSoftKeyboard() {
        (activity!!.getSystemService(Activity.INPUT_METHOD_SERVICE) as InputMethodManager).also {
            it.toggleSoftInput(InputMethodManager.HIDE_IMPLICIT_ONLY, 0)
        }
    }

    override fun showBottomSheetDialog() {
        cartBottomSheetDialogFragment.show(childFragmentManager,
                cartBottomSheetDialogFragment.tag)
    }

    override fun setCartItem(cart: MutableList<Ingredient>) {
        cartBottomSheetDialogFragment.cart = cart.toMutableList()
    }

    override fun hideSearchSuggestion() {
        rv_ingredients.visibility = View.GONE
    }

    @NeedsPermission(Manifest.permission.CAMERA)
    fun startCameraIntent() {
        var photoFile: File? = null
        try {
            photoFile = createImageFile()
        } catch (e: IOException) {
            e.printStackTrace()
        }

        if (photoFile != null) {
            val photoURI = FileProvider.getUriForFile(
                    activity!!,
                    Constants.PROVIDER_PACKAGE_NAME,
                    photoFile)
            val intent = Intent(MediaStore.ACTION_IMAGE_CAPTURE)
            intent.putExtra(MediaStore.EXTRA_OUTPUT, photoURI)
            startActivityForResult(intent, CAMERA_REQUEST_CODE)
        }
    }

    @Throws(IOException::class)
    private fun createImageFile(): File {
        val timeStamp = SimpleDateFormat(Constants.YYYY_MM_DD_FORMAT, Locale.US).format(Date())
        val storageDir = activity?.getExternalFilesDir(Environment.DIRECTORY_PICTURES)
        val image = File.createTempFile(timeStamp, Constants.JPG, storageDir)

        currentPhotoPath = image.absolutePath
        return image
    }

    private fun showPopupMenu(view: View) {
        PopupMenu(activity, view).also {
            it.menuInflater.inflate(R.menu.search_ingredient_menu, it.menu)
            it.setOnMenuItemClickListener { item ->
                when (item.itemId) {
                    R.id.menu_signout -> {
                        FirebaseAuth.getInstance().signOut()
                        fragmentListener?.navigateUserAccountActivity()
                        true
                    }
                    else -> false
                }
            }
            it.show()
        }
    }

    internal interface FragmentListener {
        fun navigateUserAccountActivity()
        fun navigateRecipeListActivity(ingredients: MutableList<Ingredient>)
    }

    companion object {
        private const val AUTOCOMPLETE_DELAY: Int = 600
        private const val CAMERA_REQUEST_CODE: Int = 101
        private const val BUNDLE_CART: String = "BUNDLE_CART"

        @JvmStatic
        fun newInstance(): IngredientsFragment = IngredientsFragment()
    }
}